feat(typeahead): option to configure accent sensitivity in highlight

fredpeaks opened the pull request:
Before submitting a pull request, please make sure you have at least performed the following:

 - [x] read and followed the [CONTRIBUTING.md](https://github.com/ng-bootstrap/ng-bootstrap/blob/master/CONTRIBUTING.md) guide.
 - [x] built and tested the changes locally.
 - [x] added/updated any applicable tests.
 - [x] added/updated any applicable API documentation.
 - [x] added/updated any applicable demos.

fredpeaks commented:
The Travis CI build fails on IE because of the `String.normalize` function which is not present in ES5. I've added a polyfill (unorm) via yarn and added it in all the `polyfills.ts` files in the project but the build keeps failing. I suppose that the tests don't use the polyfills at all.
Is it possible to make the tests pass on IE ? Or should I add a notice in the doc to inform users that this functionality is not for IE browsers ? Or that they should include a polyfill in their `polyfills.ts` file ?
Thanks for your help.

benouat commented:
Hey @fredpeaks thanks for the contribution.
As a library to be used inside application, we can't force our user to include polyfills because of us.
We don't have any polypill dependency right now (the same way Angular does) and we would like to remain like that as much as possible.

Is there any chance you could transform the usage of `unorm` into maybe a kind of gigantic RegExp ?

fredpeaks commented:
Hey @benouat. Thanks for the answer.
I understand why you don't want to include a polyfill in the library. Unfortunately the `unorm` library is a bit complex and cannot be transformed into a RexExp or something like that... The only (bad) solution I see to make this work would be to include the `unorm` code into ng-bootstrap.
That's why I asked if it was possible to add a notice in the documentation to inform users that they should add the polyfill in their app if they intend to be compatible with IE.
I will keep looking for another solution that avoids the use of the `String.normalize` function.

fredpeaks commented:
I removed the `unorm` polyfill and replaced it by some code in `src/util/util.ts`. So browsers that implement the `String.normalize` function use it to remove accents of a string and the others use the code I added (which is a bit more "old school" but does the job). So no more polyfill and now the Travis build passes.

fbasso commented:
Hi @fredpeaks 

My two cents about it.

I had a quick look at the code, and my feeling is this is a lot of code for a functionality which will used (maybe) only in few cases. I mean, it's an important feature, but some apps can support only the most recent browsers and in this case string.normalize will be a great benefit for them.

So my idea: is it possible to implement this feature only when String.normalize is available ? This point can be explain in the documentation. And for the ones who want to have it in internet explorer, up to them to polyfill String.normalize with whatever they want.

@benouat @maxokorokov @fredpeaks, what do you think ?

benouat commented:
> So my idea: is it possible to implement this feature only when String.normalize is available ? This point can be explain in the documentation. And for the ones who want to have it in internet explorer, up to them to polyfill String.normalize with whatever they want.

I like that idea

fredpeaks commented:
I'm okay with this idea, that's what I asked in my two first comments.
In addition to the documentation notice I could use `console.error` with a message saying that a polyfill should be added if the feature is used in a browser where String.normalize isn't available.  Is it okay for you or do I just add the documentation notice ?

fbasso commented:
Not sure about the console.error. We can just push any notification, but I'm afraid it will result in issues opened incorrectly. It's just a detail, but maybe a console.warn is more appropriate. 

The tests need to be adapted depending on the browser.

And the documentation is mandatory for sure.

benouat commented:
> I'm okay with this idea, that's what I asked in my two first comments.

My bad, I missed it in the first place.

> a console.warn is more appropriate

I approve that one. We can't fire console.error like that. Another option (I have actually no clue if doable) would be to setup a tslint warning when simply using `String.normalize` to get the warning at save/build time.

> The tests need to be adapted depending on the browser

For unit test, we can test that it's stripping properly thing in chrome.
Then, I think we will have to rely on e2e tests here. We need a polyfill to test only on SauceLabs with IE. So it has to be done in the e2e test app.

fbasso commented:
Agree. It's more usual to always use the normalization (if available, + doc). But is it a breaking change in this case ? (or can we consider it's a fix ?)

fredpeaks commented:
@benouat @maxokorokov @fbasso I think the input makes sense because it avoids breaking change and it offers more flexibility. I'm sure there is a use case where someone will not want the highlight to be accent insensitive. I agree it's a rare situation and I think the default value should be `true` but I didn't want to introduce a breaking change so I've made it `false`.
About the tests, I've made different tests depending on the presence of `String.normalize` so it works with or without a polyfill. But I can do some changes if needed.
@benouat About the tslint warning : do you suggest that a tslint warning is always shown at build time or just depending on the targeted browsers or ES version ?

benouat commented:
> @benouat About the tslint warning : do you suggest that a tslint warning is always shown at build time or just depending on the targeted browsers or ES version ?

Forget about that! It won't fly, as the call to `String::normalize` will be in lib code, and not written by the user themselves

maxokorokov reviewed (approved):
LGTM, just rebased and squashed the commits
